server/commands/list_objects.go

```diff
@@ -0,0 +1,370 @@
+package commands
+
+import (
+	"context"
+	"sort"
+	"sync/atomic"
+	"time"
+
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/utils"
+	serverErrors "github.com/openfga/openfga/server/errors"
+	"github.com/openfga/openfga/storage"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	MaximumConcurrentChecks = 100 // todo(jon-whit): make this configurable, but for now limit to 100 concurrent checks
+)
+
+type ListObjectsQuery struct {
+	datastore             storage.OpenFGADatastore
+	logger                logger.Logger
+	tracer                trace.Tracer
+	meter                 metric.Meter
+	ListObjectsDeadline   time.Duration
+	ListObjectsMaxResults uint32
+	ResolveNodeLimit      uint32
+}
+
+// NewListObjectsQuery creates a ListObjectsQuery
+func NewListObjectsQuery(datastore storage.OpenFGADatastore, tracer trace.Tracer, logger logger.Logger, meter metric.Meter, listObjectsDeadline time.Duration, listObjectsMaxResults uint32, ResolveNodeLimit uint32) *ListObjectsQuery {
+	return &ListObjectsQuery{
+		datastore:             datastore,
+		logger:                logger,
+		tracer:                tracer,
+		meter:                 meter,
+		ListObjectsDeadline:   listObjectsDeadline,
+		ListObjectsMaxResults: listObjectsMaxResults,
+		ResolveNodeLimit:      ResolveNodeLimit,
+	}
+}
+
+// Execute the ListObjectsQuery, returning a list of object IDs
+func (q *ListObjectsQuery) Execute(ctx context.Context, req *openfgapb.ListObjectsRequest) (*openfgapb.ListObjectsResponse, error) {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return nil, err
+	}
+	iter, err := q.getTupleIterator(ctx, req.StoreId, req.Type)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Stop()
+
+	g := new(errgroup.Group)
+	g.SetLimit(MaximumConcurrentChecks)
+
+	uniqueObjectIdsSet := utils.NewConcurrentSet()
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	var cancel context.CancelFunc
+	if q.ListObjectsDeadline != 0 {
+		timeoutCtx, cancel = context.WithTimeout(ctx, q.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	objectIDs := make([]string, 0)
+
+	go func() {
+		var objectsFound uint32
```

Comment:
What's the reasoning for using `uint32` instead of `int32`?

Reply:
🤷‍♀️  does it make a difference?

Reply:
That's why I'm asking the selection. `atomic.LoadUint32(&objectsFound)` as opposed `objectsFound >=`, just making sure that we don't mess up in the future.

Reply:
`atomic.LoadUint32` ensures that we don't mess up the counter from different goroutines